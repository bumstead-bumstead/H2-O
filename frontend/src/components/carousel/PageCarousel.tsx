import type { PropsWithChildren, RefObject } from 'react';
import { createContext, memo, useRef } from 'react';
import styled from '@emotion/styled';

interface CarouselProps {
  currentSlide: number;
}
function Carousel({ children, currentSlide }: PropsWithChildren<CarouselProps>) {
  return (
    <CarouselContainer data-testid='carousel'>
      <CarouselSlideList currentSlide={currentSlide}>{children}</CarouselSlideList>
    </CarouselContainer>
  );
}

type PageContextType = RefObject<HTMLLIElement>;

export const PageContext = createContext<PageContextType | null>(null);

function Slide({ children }: PropsWithChildren) {
  const pageRef = useRef<HTMLLIElement | null>(null);

  return (
    <PageContext.Provider value={pageRef}>
      <CarouselSlide ref={pageRef}>{children}</CarouselSlide>
    </PageContext.Provider>
  );
}

Slide.displayName = 'Slide';

Carousel.Slide = memo(Slide);
export default Carousel;

const CarouselContainer = styled.div`
  background-color: ${({ theme }) => theme.colors.blueBg};
  width: 100%;
  overflow-x: hidden;
`;

const CarouselSlideList = styled.ul<Pick<CarouselProps, 'currentSlide'>>`
  display: flex;
  gap: 12px;
  height: calc(100vh - 60px);
  transition: transform 1s ease-in-out;
  transform: translateX(calc(-${({ currentSlide }) => currentSlide} * (100% + 12px)));
`;

const CarouselSlide = styled.li`
  position: relative;
  width: 100%;
  min-height: calc(100vh - 60px);
  flex-shrink: 0;
  transform: translateX(0);
  overflow-x: hidden;
  overflow-y: scroll;
  -webkit-transform: translate3d(0, 0, 0);
`;
