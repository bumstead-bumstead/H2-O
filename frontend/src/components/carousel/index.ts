export { default as PageCarousel } from './PageCarousel';
