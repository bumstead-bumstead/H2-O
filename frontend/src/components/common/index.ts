export * from './Flex';
export * from './Typography';
export * from './Banner';
export * from './Button';
export * from './Card';
export * from './Dimmed';
export * from './Divider';
export * from './ErrorBoundary';
export * from './Footer';
export * from './HashTag';
export * from './Header';
export * from './HMGTag';
export * from './Icon';
export * from './MainSelector';
export * from './Loading';
export * from './PriceStaticBar';
export * from './Portal';
export * from './Popup';
export * from './Circle';
export * from './Toggle';
