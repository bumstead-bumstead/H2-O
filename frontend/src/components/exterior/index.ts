export { default as ExteriorCarImg } from './ExteriorCarImg';
export * from './ExteriorCard';
export { default as ExteriorSelector } from './ExteriorSelector';
