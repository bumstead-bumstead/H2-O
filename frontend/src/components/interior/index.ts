export { default as InteriorSelector } from './InteriorSelector';
export * from './InteriorCard';
