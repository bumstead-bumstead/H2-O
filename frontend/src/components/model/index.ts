export { default as ModelOptionDetail } from './ModelOptionDetail';
export { default as ModelTypeSelector } from './ModelTypeSelector';
export { default as ModelTypeCard } from './ModelTypeOptionList';
export { default as ModelTypeOptionList } from './ModelTypeOptionList';
export * from './BottomHMGData';
