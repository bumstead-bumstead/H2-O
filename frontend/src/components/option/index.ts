export * from './OptionBanner';
export { default as DefaultOptionSelector } from './DefaultOptionSelector';
export { default as ExtraOptionSelector } from './ExtraOptionSelector';
