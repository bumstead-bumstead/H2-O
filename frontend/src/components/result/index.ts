export { default as ResultBanner } from './ResultBanner';
export { default as DetailEstimate } from './DetailEstimate';
export { default as EstimateAccordian } from './EstimateAccordian';
export { default as ResultFooter } from './ResultFooter';
