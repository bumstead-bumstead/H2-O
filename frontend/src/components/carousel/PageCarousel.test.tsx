import { render, screen } from '@/tests/test-util';
import { PageCarousel as Carousel } from '@/components/carousel';

describe('PageCarousel 컴포넌트 테스트', () => {
  it('PageCarousel이 정상적으로 렌더링 된다.', () => {
    render(
      <Carousel currentSlide={1}>
        <Carousel.Slide>페이지1</Carousel.Slide>
        <Carousel.Slide>페이지2</Carousel.Slide>
        <Carousel.Slide>페이지3</Carousel.Slide>
      </Carousel>,
    );

    expect(screen.getByTestId('carousel')).toBeInTheDocument();
  });

  it('PageCarousel내 페이지 슬라이드가 정상적으로 렌더링 된다.', () => {
    render(
      <Carousel currentSlide={1}>
        <Carousel.Slide>페이지1</Carousel.Slide>
        <Carousel.Slide>페이지2</Carousel.Slide>
        <Carousel.Slide>페이지3</Carousel.Slide>
      </Carousel>,
    );

    expect(screen.getByText('페이지1')).toBeInTheDocument();
    expect(screen.getByText('페이지2')).toBeInTheDocument();
    expect(screen.getByText('페이지3')).toBeInTheDocument();
  });

  it('PageCarousel 스냅샷 테스트', () => {
    render(
      <Carousel currentSlide={1}>
        <Carousel.Slide>페이지1</Carousel.Slide>
        <Carousel.Slide>페이지2</Carousel.Slide>
        <Carousel.Slide>페이지3</Carousel.Slide>
      </Carousel>,
    );

    expect(screen.getByTestId('carousel')).toMatchSnapshot();
  });
});
