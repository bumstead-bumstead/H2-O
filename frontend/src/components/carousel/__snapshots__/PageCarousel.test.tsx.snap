// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageCarousel 컴포넌트 테스트 > PageCarousel 스냅샷 테스트 1`] = `
.emotion-0 {
  background-color: #F4F5F7;
  width: 100%;
  overflow-x: hidden;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 12px;
  height: calc(100vh - 60px);
  -webkit-transition: -webkit-transform 1s ease-in-out;
  transition: transform 1s ease-in-out;
  -webkit-transform: translateX(calc(-1 * (100% + 12px)));
  -moz-transform: translateX(calc(-1 * (100% + 12px)));
  -ms-transform: translateX(calc(-1 * (100% + 12px)));
  transform: translateX(calc(-1 * (100% + 12px)));
}

.emotion-2 {
  position: relative;
  width: 100%;
  min-height: calc(100vh - 60px);
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transform: translateX(0);
  -moz-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  overflow-x: hidden;
  overflow-y: scroll;
  -webkit-transform: translate3d(0, 0, 0);
}

<div
  class="emotion-0"
  data-testid="carousel"
>
  <ul
    class="emotion-1"
  >
    <li
      class="emotion-2"
    >
      페이지1
    </li>
    <li
      class="emotion-2"
    >
      페이지2
    </li>
    <li
      class="emotion-2"
    >
      페이지3
    </li>
  </ul>
</div>
`;
